state_machine/states/impl/waypoint_impl.py

```diff
@@ -52,98 +52,111 @@ async def run(self: Waypoint) -> State:
+    boundary_points: list[BoundarylistUtm] = gps_dict["boundary_points_utm"]
+    boundary_points.pop()  # The last point is a duplicate of the first
 
-            goto_points.pop()  # The last point is just the waypoint
+    boundary_vertices: list[Point] = []
+    for point in boundary_points:
+        boundary_vertices.append(Point(point.easting, point.northing))
 
-            lat_deg: float
-            lon_deg: float
-            lat_deg, lon_deg = utm.to_latlon(
-                waypoint.easting, waypoint.northing, waypoint.zone_number, waypoint.zone_letter
-            )
+    search_graph: list[GraphNode[Point, float]] = pathfinding.create_pathfinding_graph(
+        boundary_vertices, BOUNDARY_SHRINKAGE
+    )
 
-            logging.info("Moving to waypoint %d (lat=%f, lon=%f)", waypoint_num, lat_deg, lon_deg)
+    for waypoint_num, waypoint in enumerate(waypoints_utm):
+        drone_position: mavsdk.telemetry.Position = await anext(
+            self.drone.system.telemetry.position()
+        )
+        drone_northing, drone_easting, _, _ = utm.from_latlon(
+            drone_position.latitude_deg,
+            drone_position.longitude_deg,
+            boundary_points[0].zone_number,
+            boundary_points[0].zone_letter,
+        )
 
-            for line_segment in LineSegment.from_points(goto_points, False):
-                lat_deg, lon_deg = utm.to_latlon(
-                    line_segment.p_2.x,
-                    line_segment.p_2.y,
-                    boundary_points[0].zone_number,
-                    boundary_points[0].zone_letter,
+        goto_points: list[Point]
+        try:
+            goto_points = list(
+                pathfinding.shortest_path_between(
+                    Point(drone_easting, drone_northing),
+                    Point(waypoint.easting, waypoint.northing),
+                    search_graph,
                 )
+            )
+        except RuntimeError:
+            # Unable to find path that doesn't intersect boundary
+            # This should never happen
+            goto_points = [Point(waypoint.easting, waypoint.northing)]
+
+        path_length: float = (
+            sum(
+                line_segment.length()
+                for line_segment in LineSegment.from_points(goto_points, False)
+            )
+            + (goto_points[0] - Point(drone_easting, drone_northing)).distance_from_origin()
+        )
 
-                # Gradually move toward goal altitude
-                curr_altitude += altitude_slope * line_segment.length()
+        curr_altitude: float = drone_position.relative_altitude_m
+        altitude_slope: float = (waypoint.altitude - curr_altitude) / path_length
 
-                try:
-                    await move_to(self.drone.system, lat_deg, lon_deg, curr_altitude, 0.83)
-                except ActionError:
-                    logging.warning(ActionError)
+        goto_points.pop()  # The last point is just the waypoint
```

Comment:
We had problems with the drone not going to all waypoints during flight days, is this line necessary?

Reply:
Yes. This is simply the last point in the path to the next waypoint, which is always the waypoint. The code moves to the waypoint with more precision than the intermediate points in the path. Also, the only change to this file in the pull request was moving some waypoint code to a separate function—the waypoint code was otherwise not changed.